### PR TITLE
create library for person and pack in frontdesk

### PR DESCRIPTION
**Pull request by killianherbunot**

### CODE REVIEW

File: src/Client/Impl/ClientImpl.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function post($resource, $params)
+    {
+        $response = $this->guzzle->post(
+            $resource,
+            ['form_params' => $params, 'headers' => ['Authorization' => "Bearer $this->token"]]
```
Comment (romainkuzniak) [inline src/Client/Impl/ClientImpl.php]: Put auth in default client headers

File: src/Client/Impl/ClientFactoryImpl.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+namespace OpenClassrooms\FrontDesk\Client\Impl;
+
+use OpenClassrooms\FrontDesk\Client\ClientFactory;
+
+/**
+ * @author Killian Herbunot <killian.herbunot@openclassrooms.com>
+ */
+class ClientFactoryImpl implements ClientFactory
+{
+    public function create($key, $secret)
```
Comment (romainkuzniak) [inline src/Client/Impl/ClientFactoryImpl.php]: inherit doc

File: src/Client/Client.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+namespace OpenClassrooms\FrontDesk\Client;
+
+/**
+ * @author Killian Herbunot <killian.herbunot@openclassrooms.com>
+ */
+interface Client
+{
+    /**
+     * @param string $resource
+     * @param array $params
```
Comment (romainkuzniak) [inline src/Client/Client.php]: array what ?

File: src/Models/Impl/PackBuilderImpl.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        $this->pack->setPersonIds($personIds);
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function withStartDate($startDate)
```
Comment (romainkuzniak) [inline src/Models/Impl/PackBuilderImpl.php]: \DateTime

File: src/Models/Impl/PackBuilderImpl.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        $this->pack->setCount($count);
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function withEndDate($endDate)
```
Comment (romainkuzniak) [inline src/Models/Impl/PackBuilderImpl.php]: \DateTime + = null

File: src/Models/PackBuilder.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+     * @return PersonBuilder
+     */
+    public function withCount($count);
+
+    /**
+     * @param \DateTime $endDate
+     *
+     * @return PersonBuilder
+     */
+    public function withEndDate($endDate);
```
Comment (romainkuzniak) [inline src/Models/PackBuilder.php]: \DateTime  + = null

File: src/Models/PackBuilder.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+     * @return PersonBuilder
+     */
+    public function withPersonIds($personIds);
+
+    /**
+     * @param \DateTime $startDate
+     *
+     * @return PersonBuilder
+     */
+    public function withStartDate($startDate);
```
Comment (romainkuzniak) [inline src/Models/PackBuilder.php]: \DateTime

File: src/Models/Person.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * @return \DateTime
+     */
+    public function getJoinedAt()
+    {
+        return $this->joinedAt;
+    }
+
+    public function setJoinedAt($joinedAt)
```
Comment (romainkuzniak) [inline src/Models/Person.php]: \DateTime

File: tests/Client/Impl/ClientFactoryImplTest.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+    /**
+     * @test
+     */
+    public function createClient()
+    {
+        $clientFactory = new ClientFactoryImpl();
+        $client = $clientFactory->create(self::KEY, self::TOKEN);
+
+        $this->assertInstanceOf(ClientImpl::class, $client);
```
Comment (romainkuzniak) [inline tests/Client/Impl/ClientFactoryImplTest.php]: Not enough

File: tests/Client/Impl/ClientImplTest.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * @var ClientImpl
+     */
+    private $client;
+
+    /**
+     * @test
+     */
+    public function Construct()
```
Comment (romainkuzniak) [inline tests/Client/Impl/ClientImplTest.php]: in test factory

File: tests/Client/Impl/ClientImplTest.php [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        $guzzle = $guzzleProperty->getValue($this->client);
+        $this->assertEquals('https://' . self::KEY . '.frontdeskhq.com/api/v2/', $guzzle->getConfig('base_uri'));
+    }
+
+    /**
+     * @test
+     */
+    public function postResource()
+    {
+        $streamMock = $this->createMock(StreamInterface::class);
```
Comment (romainkuzniak) [inline tests/Client/Impl/ClientImplTest.php]: in setup

File: tests/Doubles/Models/PackStub1.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    const ID = 1;
+
+    const PERSON_IDS = ['Pack Stub 1 person ids'];
+
+    const START_DATE = '2016-09-30 11:30:00';
+
+
+    protected $count = self::COUNT;
+
+    protected $endDate = self::END_DATE;
```
Comment (romainkuzniak) [inline tests/Doubles/Models/PackStub1.php]: init in constructor

File: tests/Doubles/Models/PackStub1.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+    protected $id = self::ID;
+
+    protected $personIds = self::PERSON_IDS;
+
+    protected $startDate = self::START_DATE;
```
Comment (romainkuzniak) [inline tests/Doubles/Models/PackStub1.php]: init in constructor

File: tests/Doubles/Models/PersonStub1.php [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    const PHONE = 'Person stub 1 PHONE';
+
+    protected $address = self::ADDRESS;
+
+    protected $email = self::EMAIL;
+
+    protected $firstName = self::FIRST_NAME;
+
+    protected $joinedAt = self::JOINED_AT;
```
Comment (romainkuzniak) [inline tests/Doubles/Models/PersonStub1.php]: idem

File: tests/Repository/PackRepositoryTest.php [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+     * @var PackRepository
+     */
+    private $repository;
+
+    /**
+     * @test
+     */
+    public function createReturnId()
+    {
+        $stub = $this->createMock(Client::class);
```
Comment (romainkuzniak) [inline tests/Repository/PackRepositoryTest.php]: Build a real Client Mock

File: tests/Repository/PersonRepositoryTest.php [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+     * @var PersonRepository
+     */
+    private $repository;
+
+    /**
+     * @test
+     */
+    public function createReturnId()
+    {
+        $stub = $this->createMock(Client::class);
```
Comment (romainkuzniak) [inline tests/Repository/PersonRepositoryTest.php]: Build a real Client Mock

File: tests/Repository/PackRepositoryTest.php [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+     * @return \OpenClassrooms\FrontDesk\Models\Pack
+     */
+    private function buildPack()
+    {
+        $packBuilder = new PackBuilderImpl();
+
+        return $packBuilder
+            ->create()
+            ->withCount(PackStub1::COUNT)
+            ->withEndDate(new Carbon(PackStub1::END_DATE))
```
Comment (romainkuzniak) [inline tests/Repository/PackRepositoryTest.php]: new \DateTime

Remove Carbon Library

File: src/Models/Pack.php [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * @return \DateTime
+     */
+    public function getEndDate()
+    {
+        return $this->endDate;
+    }
+
+    /**
+     * @param \DateTime $endDate
```
Comment (romainkuzniak) [inline src/Models/Pack.php]: remove comment

File: src/Models/Pack.php [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * @return \DateTime
+     */
+    public function getStartDate()
+    {
+        return $this->startDate;
+    }
+
+    /**
+     * @param \DateTime $startDate
```
Comment (romainkuzniak) [inline src/Models/Pack.php]: remove

File: src/Models/PackBuilder.php [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * @param integer $count
+     *
+     * @return PersonBuilder
+     */
+    public function withCount($count);
+
+    /**
+     * @param \DateTime $endDate
```
Comment (romainkuzniak) [inline src/Models/PackBuilder.php]: remove

File: src/Models/PersonBuilder.php [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * @param string $firstName
+     *
+     * @return PersonBuilder
+     */
+    public function withFirstName($firstName);
+
+    /**
+     * @param \DateTime $joinedAt
```
Comment (romainkuzniak) [inline src/Models/PersonBuilder.php]: remove

File: tests/travis.ini [hunk 22]
```diff
@@ -0,0 +1,5 @@
+extension="memcache.so"
```
Comment (romainkuzniak) [inline tests/travis.ini]: remove

File: .travis.yml [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+
+php:
+  - "5.6"
+  - "7.0"
+
+before_script:
+  - composer self-update
+  - composer install --no-scripts --no-interaction
+  - composer dump-autoload -o
+  - phpenv config-add ./tests/travis.ini
```
Comment (romainkuzniak) [inline .travis.yml]: remove

File: src/Models/Pack.php [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+    public function setStartDate(\DateTime $startDate = null)
+    {
+        $this->startDate = $startDate;
+    }
+
+    /**
+     * @return array
+     */
+    public function toArray()
```
Comment (kletord) [inline src/Models/Pack.php]: Should be move into the Impl, It's not part of the public api.

File: src/Models/Person.php [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    public function setPhone($phone)
+    {
+        $this->phone = $phone;
+    }
+
+    /**
+     * @return array
+     */
+    public function toArray()
```
Comment (kletord) [inline src/Models/Person.php]: Should be move into the Impl, It's not part of the public api.

**Pull request merged**